From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: John Merchant <john@johnmerchant.dev>
Date: Wed, 28 Oct 2026 13:12:22 +0000
Subject: [PATCH 10/11] hw/arm: Add the i.MX21 SoC

An ARM926EJ-S (r0p4) with the i.MX21's interrupt controller, timers,
UARTs (the existing i.MX serial model), clocks, GPIO, CSPI, LCD and NAND
flash controllers, PWM and USB OTG registers. Peripheral and memory
controller registers that are not modelled read back what was written.

Signed-off-by: John Merchant <john@johnmerchant.dev>
---
 hw/arm/Kconfig             |   5 +
 hw/arm/fsl-imx21.c         | 183 +++++++++++++++++++++++++++++++++++++
 hw/arm/meson.build         |   1 +
 include/hw/arm/fsl-imx21.h |  94 +++++++++++++++++++
 4 files changed, 283 insertions(+)
 create mode 100644 hw/arm/fsl-imx21.c
 create mode 100644 include/hw/arm/fsl-imx21.h

diff --git a/hw/arm/Kconfig b/hw/arm/Kconfig
index 6f5b2add6..cc3afd871 100644
--- a/hw/arm/Kconfig
+++ b/hw/arm/Kconfig
@@ -489,6 +489,11 @@ config NPCM8XX
     select UNIMP
 
 
+config FSL_IMX21
+    bool
+    select IMX
+    select SSI
+
 config FSL_IMX25
     bool
     default y
diff --git a/hw/arm/fsl-imx21.c b/hw/arm/fsl-imx21.c
new file mode 100644
index 000000000..e8c27b87e
--- /dev/null
+++ b/hw/arm/fsl-imx21.c
@@ -0,0 +1,183 @@
+/*
+ * Freescale i.MX21 SoC
+ *
+ * Copyright (c) 2026 John Merchant <john@johnmerchant.dev>
+ *
+ * SPDX-License-Identifier: GPL-2.0-or-later
+ *
+ * An ARM926EJ-S with the i.MX21's interrupt controller, timers, UARTs,
+ * clocks, GPIO, SPI, LCD and NAND flash controllers. Peripheral and
+ * memory controller registers that are not modelled (DMA, SSI audio,
+ * keypad, watchdog, SDRAM and WEIM controllers and others) read back
+ * what was last written.
+ */
+
+#include "qemu/osdep.h"
+#include "qapi/error.h"
+#include "hw/arm/fsl-imx21.h"
+#include "hw/core/qdev-properties.h"
+#include "hw/core/qdev-properties-system.h"
+#include "hw/core/sysbus.h"
+#include "chardev/char.h"
+#include "system/address-spaces.h"
+#include "system/system.h"
+#include "target/arm/cpu-qom.h"
+
+static void fsl_imx21_init(Object *obj)
+{
+    FslIMX21State *s = FSL_IMX21(obj);
+    int i;
+
+    object_initialize_child(obj, "cpu", &s->cpu, ARM_CPU_TYPE_NAME("arm926"));
+    object_initialize_child(obj, "aitc", &s->aitc, TYPE_IMX21_AITC);
+    for (i = 0; i < FSL_IMX21_NUM_GPTS; i++) {
+        object_initialize_child(obj, "gpt[*]", &s->gpt[i], TYPE_IMX21_GPT);
+    }
+    for (i = 0; i < FSL_IMX21_NUM_UARTS; i++) {
+        object_initialize_child(obj, "uart[*]", &s->uart[i], TYPE_IMX_SERIAL);
+    }
+    object_initialize_child(obj, "crm", &s->crm, TYPE_IMX21_CRM);
+    object_initialize_child(obj, "gpio", &s->gpio, TYPE_IMX21_GPIO);
+    for (i = 0; i < FSL_IMX21_NUM_CSPIS; i++) {
+        object_initialize_child(obj, "cspi[*]", &s->cspi[i], TYPE_IMX21_CSPI);
+    }
+    object_initialize_child(obj, "lcdc", &s->lcdc, TYPE_IMX21_LCDC);
+    object_initialize_child(obj, "nfc", &s->nfc, TYPE_IMX21_NFC);
+}
+
+static bool fsl_imx21_map(DeviceState *dev, SysBusDevice *sbd, hwaddr addr,
+                          int irq, Error **errp)
+{
+    if (!sysbus_realize(sbd, errp)) {
+        return false;
+    }
+    sysbus_mmio_map(sbd, 0, addr);
+    if (irq >= 0) {
+        sysbus_connect_irq(sbd, 0, qdev_get_gpio_in(dev, irq));
+    }
+    return true;
+}
+
+static void fsl_imx21_realize(DeviceState *dev, Error **errp)
+{
+    FslIMX21State *s = FSL_IMX21(dev);
+    DeviceState *aitc = DEVICE(&s->aitc);
+    MemoryRegion *sysmem = get_system_memory();
+    int i;
+
+    static const struct {
+        hwaddr addr;
+        int irq;
+    } gpts[FSL_IMX21_NUM_GPTS] = {
+        { FSL_IMX21_GPT1_ADDR, FSL_IMX21_GPT1_IRQ },
+        { FSL_IMX21_GPT2_ADDR, FSL_IMX21_GPT2_IRQ },
+        { FSL_IMX21_GPT3_ADDR, FSL_IMX21_GPT3_IRQ },
+    }, uarts[FSL_IMX21_NUM_UARTS] = {
+        { FSL_IMX21_UART1_ADDR, FSL_IMX21_UART1_IRQ },
+        { FSL_IMX21_UART2_ADDR, FSL_IMX21_UART2_IRQ },
+        { FSL_IMX21_UART3_ADDR, FSL_IMX21_UART3_IRQ },
+        { FSL_IMX21_UART4_ADDR, FSL_IMX21_UART4_IRQ },
+    }, cspis[FSL_IMX21_NUM_CSPIS] = {
+        { FSL_IMX21_CSPI1_ADDR, FSL_IMX21_CSPI1_IRQ },
+        { FSL_IMX21_CSPI2_ADDR, FSL_IMX21_CSPI2_IRQ },
+        { FSL_IMX21_CSPI3_ADDR, FSL_IMX21_CSPI3_IRQ },
+    };
+
+    /* The i.MX21's core is an ARM926EJ-S r0p4. */
+    if (!object_property_set_int(OBJECT(&s->cpu), "midr", FSL_IMX21_MIDR,
+                                 errp) ||
+        !qdev_realize(DEVICE(&s->cpu), NULL, errp)) {
+        return;
+    }
+
+    if (!sysbus_realize(SYS_BUS_DEVICE(aitc), errp)) {
+        return;
+    }
+    sysbus_mmio_map(SYS_BUS_DEVICE(aitc), 0, FSL_IMX21_AITC_ADDR);
+    sysbus_connect_irq(SYS_BUS_DEVICE(aitc), 0,
+                       qdev_get_gpio_in(DEVICE(&s->cpu), ARM_CPU_IRQ));
+    sysbus_connect_irq(SYS_BUS_DEVICE(aitc), 1,
+                       qdev_get_gpio_in(DEVICE(&s->cpu), ARM_CPU_FIQ));
+
+    for (i = 0; i < FSL_IMX21_NUM_GPTS; i++) {
+        if (!fsl_imx21_map(aitc, SYS_BUS_DEVICE(&s->gpt[i]), gpts[i].addr,
+                           gpts[i].irq, errp)) {
+            return;
+        }
+    }
+
+    for (i = 0; i < FSL_IMX21_NUM_UARTS; i++) {
+        DeviceState *uart = DEVICE(&s->uart[i]);
+        g_autofree char *chr = object_property_get_str(OBJECT(uart),
+                                                       "chardev", NULL);
+
+        /* A board may have attached something else already. */
+        if (!chr || !*chr) {
+            qdev_prop_set_chr(uart, "chardev", serial_hd(i));
+        }
+        if (!fsl_imx21_map(aitc, SYS_BUS_DEVICE(uart), uarts[i].addr,
+                           uarts[i].irq, errp)) {
+            return;
+        }
+    }
+
+    if (!fsl_imx21_map(aitc, SYS_BUS_DEVICE(&s->crm), FSL_IMX21_CRM_ADDR,
+                       -1, errp) ||
+        !fsl_imx21_map(aitc, SYS_BUS_DEVICE(&s->gpio), FSL_IMX21_GPIO_ADDR,
+                       FSL_IMX21_GPIO_IRQ, errp) ||
+        !fsl_imx21_map(aitc, SYS_BUS_DEVICE(&s->lcdc), FSL_IMX21_LCDC_ADDR,
+                       FSL_IMX21_LCDC_IRQ, errp) ||
+        !fsl_imx21_map(aitc, SYS_BUS_DEVICE(&s->nfc), FSL_IMX21_NFC_ADDR,
+                       FSL_IMX21_NFC_IRQ, errp)) {
+        return;
+    }
+
+    for (i = 0; i < FSL_IMX21_NUM_CSPIS; i++) {
+        if (!fsl_imx21_map(aitc, SYS_BUS_DEVICE(&s->cspi[i]), cspis[i].addr,
+                           cspis[i].irq, errp)) {
+            return;
+        }
+    }
+
+    s->pwm = sysbus_create_simple("imx21.pwm", FSL_IMX21_PWM_ADDR, NULL);
+    s->usbotg = sysbus_create_simple("imx21.usbotg", FSL_IMX21_USBOTG_ADDR,
+                                     qdev_get_gpio_in(aitc,
+                                                      FSL_IMX21_USBHOST_IRQ));
+
+    /* Everything else in the peripheral and memory controller spaces. */
+    if (!memory_region_init_ram(&s->aipi_regs, OBJECT(dev), "imx21.aipi",
+                                FSL_IMX21_AIPI_SIZE, errp) ||
+        !memory_region_init_ram(&s->emi_regs, OBJECT(dev), "imx21.emi",
+                                FSL_IMX21_EMI_SIZE, errp)) {
+        return;
+    }
+    memory_region_add_subregion_overlap(sysmem, FSL_IMX21_AIPI_ADDR,
+                                        &s->aipi_regs, -1000);
+    memory_region_add_subregion_overlap(sysmem, FSL_IMX21_EMI_ADDR,
+                                        &s->emi_regs, -1000);
+}
+
+static void fsl_imx21_class_init(ObjectClass *oc, const void *data)
+{
+    DeviceClass *dc = DEVICE_CLASS(oc);
+
+    dc->realize = fsl_imx21_realize;
+    dc->desc = "i.MX21 SoC";
+    /* Reason: uses serial_hd() in the realize function */
+    dc->user_creatable = false;
+}
+
+static const TypeInfo fsl_imx21_type_info = {
+    .name = TYPE_FSL_IMX21,
+    .parent = TYPE_DEVICE,
+    .instance_size = sizeof(FslIMX21State),
+    .instance_init = fsl_imx21_init,
+    .class_init = fsl_imx21_class_init,
+};
+
+static void fsl_imx21_register_types(void)
+{
+    type_register_static(&fsl_imx21_type_info);
+}
+
+type_init(fsl_imx21_register_types)
diff --git a/hw/arm/meson.build b/hw/arm/meson.build
index 4233a800b..783538cb8 100644
--- a/hw/arm/meson.build
+++ b/hw/arm/meson.build
@@ -40,6 +40,7 @@ arm_common_ss.add(when: 'CONFIG_STM32L4X5_SOC', if_true: files('stm32l4x5_soc.c'
 arm_common_ss.add(when: 'CONFIG_XLNX_ZYNQMP_ARM', if_true: files('xlnx-zynqmp.c', 'xlnx-zcu102.c'))
 arm_common_ss.add(when: 'CONFIG_XLNX_VERSAL', if_true: files('xlnx-versal.c', 'xlnx-versal-virt.c'))
 arm_common_ss.add(when: 'CONFIG_FSL_IMX25', if_true: files('fsl-imx25.c', 'imx25_pdk.c'))
+arm_common_ss.add(when: 'CONFIG_FSL_IMX21', if_true: files('fsl-imx21.c'))
 arm_common_ss.add(when: 'CONFIG_FSL_IMX31', if_true: files('fsl-imx31.c', 'kzm.c'))
 arm_common_ss.add(when: 'CONFIG_FSL_IMX6', if_true: files('fsl-imx6.c'))
 arm_common_ss.add(when: 'CONFIG_ASPEED_SOC', if_true: files(
diff --git a/include/hw/arm/fsl-imx21.h b/include/hw/arm/fsl-imx21.h
new file mode 100644
index 000000000..b79445b86
--- /dev/null
+++ b/include/hw/arm/fsl-imx21.h
@@ -0,0 +1,94 @@
+/*
+ * Freescale i.MX21 SoC
+ *
+ * Copyright (c) 2026 John Merchant <john@johnmerchant.dev>
+ *
+ * SPDX-License-Identifier: GPL-2.0-or-later
+ */
+
+#ifndef FSL_IMX21_H
+#define FSL_IMX21_H
+
+#include "hw/intc/imx21_aitc.h"
+#include "hw/timer/imx21_gpt.h"
+#include "hw/char/imx_serial.h"
+#include "hw/misc/imx21_crm.h"
+#include "hw/gpio/imx21_gpio.h"
+#include "hw/ssi/imx21_cspi.h"
+#include "hw/display/imx21_lcdc.h"
+#include "hw/block/imx21_nfc.h"
+#include "qemu/units.h"
+#include "system/memory.h"
+#include "target/arm/cpu.h"
+#include "qom/object.h"
+
+#define TYPE_FSL_IMX21 "fsl-imx21"
+OBJECT_DECLARE_SIMPLE_TYPE(FslIMX21State, FSL_IMX21)
+
+#define FSL_IMX21_MIDR 0x41069264
+
+#define FSL_IMX21_NUM_UARTS 4
+#define FSL_IMX21_NUM_GPTS 3
+#define FSL_IMX21_NUM_CSPIS 3
+
+struct FslIMX21State {
+    DeviceState parent_obj;
+
+    ARMCPU cpu;
+    IMX21AITCState aitc;
+    IMX21GPTState gpt[FSL_IMX21_NUM_GPTS];
+    IMXSerialState uart[FSL_IMX21_NUM_UARTS];
+    IMX21CRMState crm;
+    IMX21GPIOState gpio;
+    IMX21CSPIState cspi[FSL_IMX21_NUM_CSPIS];
+    IMX21LCDCState lcdc;
+    IMX21NFCState nfc;
+    DeviceState *pwm;
+    DeviceState *usbotg;
+    /* Peripheral and memory controller registers not modelled yet. */
+    MemoryRegion aipi_regs;
+    MemoryRegion emi_regs;
+};
+
+/* Memory map */
+#define FSL_IMX21_AIPI_ADDR     0x10000000
+#define FSL_IMX21_AIPI_SIZE     0x41000
+#define FSL_IMX21_GPT1_ADDR     0x10003000
+#define FSL_IMX21_GPT2_ADDR     0x10004000
+#define FSL_IMX21_GPT3_ADDR     0x10005000
+#define FSL_IMX21_PWM_ADDR      0x10006000
+#define FSL_IMX21_UART1_ADDR    0x1000a000
+#define FSL_IMX21_UART2_ADDR    0x1000b000
+#define FSL_IMX21_UART3_ADDR    0x1000c000
+#define FSL_IMX21_UART4_ADDR    0x1000d000
+#define FSL_IMX21_CSPI1_ADDR    0x1000e000
+#define FSL_IMX21_CSPI2_ADDR    0x1000f000
+#define FSL_IMX21_GPIO_ADDR     0x10015000
+#define FSL_IMX21_CSPI3_ADDR    0x10017000
+#define FSL_IMX21_LCDC_ADDR     0x10021000
+#define FSL_IMX21_USBOTG_ADDR   0x10024000
+#define FSL_IMX21_CRM_ADDR      0x10027000
+#define FSL_IMX21_AITC_ADDR     0x10040000
+#define FSL_IMX21_SDRAM0_ADDR   0xc0000000
+#define FSL_IMX21_SDRAM0_SIZE   (64 * MiB)
+#define FSL_IMX21_EMI_ADDR      0xdf000000
+#define FSL_IMX21_EMI_SIZE      0x3000
+#define FSL_IMX21_NFC_ADDR      0xdf003000
+
+/* Interrupts */
+#define FSL_IMX21_CSPI3_IRQ     6
+#define FSL_IMX21_GPIO_IRQ      8
+#define FSL_IMX21_CSPI2_IRQ     15
+#define FSL_IMX21_CSPI1_IRQ     16
+#define FSL_IMX21_UART4_IRQ     17
+#define FSL_IMX21_UART3_IRQ     18
+#define FSL_IMX21_UART2_IRQ     19
+#define FSL_IMX21_UART1_IRQ     20
+#define FSL_IMX21_GPT3_IRQ      24
+#define FSL_IMX21_GPT2_IRQ      25
+#define FSL_IMX21_GPT1_IRQ      26
+#define FSL_IMX21_NFC_IRQ       29
+#define FSL_IMX21_USBHOST_IRQ   55
+#define FSL_IMX21_LCDC_IRQ      61
+
+#endif /* FSL_IMX21_H */
